File: apps/api/app/recurrences/application/list_recurrences.py
from __future__ import annotations

from app.recurrences.domain.entities import Recurrence, RecurrenceKind
from app.recurrences.ports.repository import RecurrenceRepository


class ListRecurrences:
    def __init__(self, repository: RecurrenceRepository) -> None:
        self._repository = repository

    def __call__(self, kind: RecurrenceKind | None = None) -> list[Recurrence]:
        recurrences = self._repository.list_active()
        if kind is not None:
            recurrences = [item for item in recurrences if item.kind == kind]
        return sorted(recurrences, key=lambda item: item.id)

File: apps/api/internal/recurrences/application/list_recurrences.go
package application

import (
	"context"

	"caudal-api/internal/recurrences/domain"
	"caudal-api/internal/recurrences/ports"
)

type ListRecurrences struct {
	repository ports.Repository
}

func NewListRecurrences(repository ports.Repository) ListRecurrences {
	return ListRecurrences{repository: repository}
}

func (useCase ListRecurrences) Execute(ctx context.Context, kind *domain.Kind) ([]domain.Recurrence, error) {
	return useCase.repository.List(ctx, kind)
}
